File: internal/testdata/samples/sample.py
"""
Sample Python file for testing cross-language functionality.
Contains various Python constructs that map to universal concepts.
"""

import os
import json
from typing import Optional, Dict, Any
from dataclasses import dataclass

# Constants following Python naming convention
DATABASE_URL = "postgres://localhost:5432/mydb"
API_VERSION = "v1.2.0"

# Global variables
global_counter = 0
is_initialized = False


@dataclass
class User:
    """User represents a system user with authentication data."""
    id: int
    name: str
    email: str
    password: Optional[str] = None

    def set_password(self, password: str) -> None:
        """Set the user password."""
        self.password = password

    def get_display_name(self) -> str:
        """Return formatted display name."""
        return f"{self.name} <{self.email}>"

    def is_valid(self) -> bool:
        """Check if user data is valid."""
        return validate_user(self)


class MockUser:
    """MockUser is a test user structure."""
    
    def __init__(self, user_id: int, name: str):
        self.id = user_id
        self.name = name


def new_user(name: str, email: str) -> User:
    """Create a new user instance."""
    return User(id=0, name=name, email=email)


def test_create_user():
    """Test user creation functionality."""
    user = new_user("John Doe", "john@example.com")
    assert user.name == "John Doe", "Name not set correctly"


def test_user_email():
    """Test user email validation."""
    user = User(id=1, name="Test", email="invalid-email")
    assert validate_user(user), "Email validation failed"


def validate_user(user: User) -> bool:
    """Perform user validation."""
    return len(user.email) > 0 and len(user.name) > 0


def main():
    """Main function demonstrating user functionality."""
    user = new_user("Admin", "admin@example.com")
    user.set_password("secret123")
    
    if user.is_valid():
        print(f"User is valid: {user.get_display_name()}")
    else:
        print("Invalid user data")
        os.exit(1)


if __name__ == "__main__":
    main()

File: internal/testdata/samples/sample.ts
/**
 * Sample TypeScript file for testing cross-language functionality.
 * Contains various TypeScript constructs that map to universal concepts.
 */

import * as fs from 'fs';
import * as path from 'path';

// Type definitions
interface UserData {
    id: number;
    name: string;
    email: string;
    password?: string;
}

// Constants
const DATABASE_URL: string = "postgres://localhost:5432/mydb";
const API_VERSION: string = "v1.2.0";

// Global variables
let globalCounter: number = 0;
var isInitialized: boolean = false;

/**
 * User class represents a system user with authentication data
 */
class User implements UserData {
    public id: number;
    public name: string;
    public email: string;
    private password?: string;

    constructor(id: number, name: string, email: string) {
        this.id = id;
        this.name = name;
        this.email = email;
    }

    /**
     * Set the user password
     */
    setPassword(password: string): void {
        this.password = password;
    }

    /**
     * Return formatted display name
     */
    getDisplayName(): string {
        return `${this.name} <${this.email}>`;
    }

    /**
     * Check if user data is valid
     */
    isValid(): boolean {
        return validateUser(this);
    }

    /**
     * Get user as JSON object
     */
    toJSON(): UserData {
        return {
            id: this.id,
            name: this.name,
            email: this.email,
            password: this.password
        };
    }
}

/**
 * MockUser is a test user structure
 */
class MockUser {
    constructor(
        public readonly id: number,
        public readonly name: string
    ) {}
}

/**
 * Create a new user instance
 */
function newUser(name: string, email: string): User {
    return new User(0, name, email);
}

/**
 * Test user creation functionality
 */
function testCreateUser(): void {
    const user = newUser("John Doe", "john@example.com");
    console.assert(user.name === "John Doe", "Name not set correctly");
}

/**
 * Test user email validation
 */
function testUserEmail(): void {
    const user = new User(1, "Test", "invalid-email");
    console.assert(validateUser(user), "Email validation failed");
}

/**
 * Perform user validation
 */
function validateUser(user: UserData): boolean {
    return user.email.length > 0 && user.name.length > 0;
}

/**
 * Arrow function for user processing with type annotation
 */
const processUser = (user: User): string => {
    if (user.isValid()) {
        return user.getDisplayName();
    }
    return "Invalid user";
};

/**
 * Generic function for data processing
 */
function processData<T>(data: T[], processor: (item: T) => string): string[] {
    return data.map(processor);
}

/**
 * Enum for user roles
 */
enum UserRole {
    ADMIN = "admin",
    USER = "user",
    GUEST = "guest"
}

/**
 * Type alias for user processing function
 */
type UserProcessor = (user: User) => string;

/**
 * Main function demonstrating user functionality
 */
function main(): void {
    const user = newUser("Admin", "admin@example.com");
    user.setPassword("secret123");
    
    if (user.isValid()) {
        console.log(`User is valid: ${user.getDisplayName()}`);
    } else {
        console.log("Invalid user data");
        process.exit(1);
    }
}

// Export for testing
export { 
    User, 
    MockUser, 
    UserData, 
    UserRole, 
    UserProcessor,
    newUser, 
    validateUser, 
    processUser, 
    processData 
};

// Run main if this is the entry point
if (require.main === module) {
    main();
}